File: cl-jvm/JVMCompiler.java

import org.objectweb.asm.AnnotationVisitor;
import org.objectweb.asm.Attribute;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.ConstantDynamic;
import org.objectweb.asm.FieldVisitor;
import org.objectweb.asm.Handle;
import org.objectweb.asm.Label;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.RecordComponentVisitor;
import org.objectweb.asm.Type;
import org.objectweb.asm.TypePath;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class JVMCompiler implements Opcodes {

    private enum State {
	PREPARE,
	CODE,
	FINISH
    };

    private byte[] _data;
    private State _state;
    private String _className;
    private ClassWriter _classWriter;
    private MethodVisitor _runMV;

    public JVMCompiler()
    {
	_data = null;
	_state = State.PREPARE;
	_className = null;
	_classWriter = null;
	_runMV = null;
    }

    public void setClassName(String className) {
	_className = className;
    }

    public void begin() throws RuntimeException {
	if (_className == null) {
	    throw new RuntimeException("Invalid call to begin(): className not set");
	}
	if (_state != State.PREPARE) {
	    throw new RuntimeException("begin() has already been called");
	}

	_state = State.CODE;
	_createClassWriter();
	_createInitMethod();
	_beginRunMethod();
    }

    public void end() throws RuntimeException {
	if (_state != State.CODE) {
	    throw new RuntimeException("end() can be called only after calling begin(), and only once");
	}

	_state = State.FINISH;
	_endRunMethod();
	_finishClass();
    }

    public byte[] getBytes() throws RuntimeException {
	if (_state != State.FINISH) {
	    throw new RuntimeException("getBytes() can only be called after calling end()");
	}

	return _data;
    }

    public void saveToFile() throws RuntimeException, FileNotFoundException, IOException {
	if (_state != State.FINISH) {
	    throw new RuntimeException("saveToFile() can only be called after calling end()");
	}


	File file = new File(_className + ".class");
	FileOutputStream os = new FileOutputStream(file);
	os.write(_data);
    }



    public void i_bipush(int val)
    {
       _runMV.visitIntInsn(BIPUSH, val);
    }

    public void i_istore(int pos) {
	_runMV.visitVarInsn(ISTORE, pos);
    }

    public void i_iload(int pos) {
	_runMV.visitVarInsn(ILOAD, pos);
    }

    public void i_iadd() {
	_runMV.visitInsn(IADD);
    }

    public void i_isub() {
	_runMV.visitInsn(ISUB);
    }

    public void i_imul() {
	_runMV.visitInsn(IMUL);
    }

    public void i_idiv() {
	_runMV.visitInsn(IDIV);
    }

    public void i_irem() {
	_runMV.visitInsn(IREM);
    }

    public void i_ireturn() {
	_runMV.visitInsn(IRETURN);
    }

    public void setMaxs(int maxStack, int maxLocals)
    {
	_runMV.visitMaxs(maxStack, maxLocals);
    }


    private void _createClassWriter() {
	_classWriter = new ClassWriter(0);
	_classWriter.visit(V13, ACC_PUBLIC | ACC_SUPER, _className, null, "java/lang/Object", null);
	_classWriter.visitSource(_className + ".java", null);
    }

    private void _finishClass() {
	_classWriter.visitEnd();
	_data = _classWriter.toByteArray();
    }

    private void _createInitMethod() {
	MethodVisitor mv = _classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
	mv.visitCode();
	mv.visitVarInsn(ALOAD, 0);
	mv.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V", false);
	mv.visitInsn(RETURN);
	mv.visitMaxs(1, 1);
	mv.visitEnd();
    }

    private void _beginRunMethod() {
	_runMV = _classWriter.visitMethod(ACC_PUBLIC | ACC_STATIC, "run", "()I", null, null);
	_runMV.visitCode();
    }

    private void _endRunMethod() {
	_runMV.visitEnd();
    }






    /*
    public static byte[] dump () throws Exception {

	ClassWriter classWriter = new ClassWriter(0);
	MethodVisitor methodVisitor;

	classWriter.visit(V13, ACC_PUBLIC | ACC_SUPER, "Compute", null, "java/lang/Object", null);

	classWriter.visitSource("Compute.java", null);

	{
	    methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
	    methodVisitor.visitCode();
	    methodVisitor.visitVarInsn(ALOAD, 0);
	    methodVisitor.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V", false);
	    methodVisitor.visitInsn(RETURN);
	    methodVisitor.visitMaxs(1, 1);
	    methodVisitor.visitEnd();
	}
	
	{
	    methodVisitor = classWriter.visitMethod(ACC_PUBLIC | ACC_STATIC, "compute", "()I", null, null);
	    methodVisitor.visitCode();

	    methodVisitor.visitIntInsn(BIPUSH, 10);
	    methodVisitor.visitVarInsn(ISTORE, 0);
	    methodVisitor.visitIntInsn(BIPUSH, 5);
	    methodVisitor.visitVarInsn(ISTORE, 1);
	    methodVisitor.visitIntInsn(BIPUSH, 23);
	    methodVisitor.visitVarInsn(ISTORE, 2);
	    methodVisitor.visitIntInsn(BIPUSH, 67);
	    methodVisitor.visitVarInsn(ISTORE, 3);
	    methodVisitor.visitIntInsn(BIPUSH, 13);
	    methodVisitor.visitVarInsn(ISTORE, 4);
	    methodVisitor.visitVarInsn(ILOAD, 0);
	    methodVisitor.visitVarInsn(ILOAD, 1);
	    methodVisitor.visitVarInsn(ILOAD, 2);
	    methodVisitor.visitInsn(IMUL);
	    methodVisitor.visitVarInsn(ILOAD, 3);
	    methodVisitor.visitInsn(IDIV);
	    methodVisitor.visitVarInsn(ILOAD, 0);
	    methodVisitor.visitInsn(IREM);
	    methodVisitor.visitInsn(IADD);
	    methodVisitor.visitVarInsn(ILOAD, 4);
	    methodVisitor.visitInsn(ISUB);
	    methodVisitor.visitInsn(IRETURN);
	    
	    methodVisitor.visitMaxs(3, 5);
	    methodVisitor.visitEnd();
	}
	
	classWriter.visitEnd();
	return classWriter.toByteArray();
    }


    public static void main(String[] args)
    {

	byte[] data = null;
	try {
	    data = Compiler.dump();
	}
	catch (Exception e) {
	    System.out.println("Failed to dump code: " + e);
	    return;
	}
	    
	try
	    {
		File file = new File("Compute.class");
		FileOutputStream os = new FileOutputStream(file);
		os.write(data);
	    }
	catch (FileNotFoundException e) {
            System.out.println("File not found" + e);
        }
        catch (IOException ioe) {
            System.out.println("Exception while writing file " + ioe);
        }
	
    }
    */
    
}

File: cl-jvm/src/jvmcompiler.rs
use jni::objects::JClass;
use jni::objects::JValue;
use jni::AttachGuard;
use jni::InitArgsBuilder;
use jni::JNIVersion;
use jni::JavaVM;

pub struct JVMCompiler<'a> {
    jvm: &'a JavaVM,
    env: AttachGuard<'a>,
    class_cl: JClass<'a>,
}

impl<'a> JVMCompiler<'a> {
    pub fn build_jvm() -> JavaVM {
        let jvm_args = InitArgsBuilder::new()
            .version(JNIVersion::V8)
            .option("-Xcheck:jni")
            .option("-Djava.class.path=.:asm.jar")
            .build()
            .expect("Failed to create JVM arguments");

        JavaVM::new(jvm_args).expect("Failed to create JVM")
    }

    pub fn new(jvm: &'a JavaVM) -> Self {
        let env: AttachGuard = jvm
            .attach_current_thread()
            .expect("Failed to create JNI env");

        let class_cl = env
            .find_class("CompilerAPI")
            .expect("Failed to find class API");
        JVMCompiler { jvm, env, class_cl }
    }

    pub fn set_class_name(&self, name: &str) {
        let a_name = self
            .env
            .new_string(name)
            .expect("Failed to create className string argument");
        self.env
            .call_static_method(
                self.class_cl,
                "setClassName",
                "(Ljava/lang/String;)V",
                &[JValue::Object(a_name.into())],
            )
            .expect("Fail to call method CompilerAPI.setClassName()");
    }

    pub fn begin(&self) {
        self.env
            .call_static_method(self.class_cl, "begin", "()V", &[])
            .expect("Fail to call method CompilerAPI.begin()");
    }

    pub fn end(&self) {
        self.env
            .call_static_method(self.class_cl, "end", "()V", &[])
            .expect("Fail to call method CompilerAPI.end()");
    }

    pub fn save_to_file(&self) {
        self.env
            .call_static_method(self.class_cl, "saveToFile", "()V", &[])
            .expect("Fail to call method CompilerAPI.saveToFile()");
    }

    pub fn i_bipush(&self, val: i32) {
        self.env
            .call_static_method(self.class_cl, "i_bipush", "(I)V", &[JValue::from(val)])
            .expect("Fail to call method CompilerAPI.i_bipush()");
    }

    pub fn i_istore(&self, val: i32) {
        self.env
            .call_static_method(self.class_cl, "i_istore", "(I)V", &[JValue::from(val)])
            .expect("Fail to call method CompilerAPI.i_istore()");
    }

    pub fn i_iload(&self, val: i32) {
        self.env
            .call_static_method(self.class_cl, "i_iload", "(I)V", &[JValue::from(val)])
            .expect("Fail to call method CompilerAPI.i_iload()");
    }

    pub fn i_iadd(&self) {
        self.env
            .call_static_method(self.class_cl, "i_iadd", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_iadd()");
    }

    pub fn i_isub(&self) {
        self.env
            .call_static_method(self.class_cl, "i_isub", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_isub()");
    }

    pub fn i_imul(&self) {
        self.env
            .call_static_method(self.class_cl, "i_imul", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_imul()");
    }

    pub fn i_idiv(&self) {
        self.env
            .call_static_method(self.class_cl, "i_idiv", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_idiv()");
    }

    pub fn i_irem(&self) {
        self.env
            .call_static_method(self.class_cl, "i_irem", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_irem()");
    }

    pub fn i_ireturn(&self) {
        self.env
            .call_static_method(self.class_cl, "i_ireturn", "()V", &[])
            .expect("Fail to call method CompilerAPI.i_ireturn()");
    }

    pub fn set_maxs(&self, max_stack: i32, max_locals: i32) {
        self.env
            .call_static_method(
                self.class_cl,
                "setMaxs",
                "(II)V",
                &[JValue::from(max_stack), JValue::from(max_locals)],
            )
            .expect("Fail to call method CompilerAPI.setMaxs()");
    }
}
